onnxruntime/core/providers/cpu/math/einsum_utils/einsum_auxiliary_ops.cc

```diff
@@ -271,35 +245,212 @@ std::unique_ptr<Tensor> Diagonal(const Tensor& input, int64_t dim_1, int64_t dim
   return output;
 }
 
-// Explicit template instantiation
+}  // namespace CpuDeviceHelpers
+
+}  // namespace DeviceHelpers
+
+// This helps decide if we need to apply (and pay the cost) of a Transpose
+bool IsTransposeRequired(size_t input_rank, const std::vector<size_t>& permutation) {
+  ORT_ENFORCE(input_rank == permutation.size(), "The rank of the input must match permutation size for Transpose");
+
+  // No transpose required for scalars
+  if (input_rank == 0) {
+    return false;
+  }
+
+  // Weeds out cases where permutation is something like [0, 1, 2] for a 3D input and so on
+  bool transpose_required = false;
+  for (size_t i = 0; i < input_rank; ++i) {
+    if (permutation[i] != i) {
+      transpose_required = true;
+      break;
+    }
+  }
+
+  return transpose_required;
+}
+
+// The following are thin wrappers over device specific helpers
+std::unique_ptr<Tensor> Transpose(const Tensor& input, const std::vector<int64_t>& input_shape_override,
+                                  const std::vector<size_t>& permutation, AllocatorPtr allocator,
+                                  void* cublas_handle, const DeviceHelpers::Transpose& device_transpose_func) {
+  auto input_rank = input_shape_override.size();
+  ORT_ENFORCE(input_rank == permutation.size(), "Length of permutation must match the rank of the input to be permutated");
+
+  std::vector<int64_t> output_dims;
+  output_dims.reserve(input_rank);
+
+  for (const auto& dim : permutation) {
+    output_dims.push_back(input_shape_override.at(dim));
```

Comment:
>input_shape_override.at(dim)); [](start = 26, length = 30)

nit: as you already checked the sizes match you don't need the overhead of the bounds check in 'at()'

Reply:
Done